### ISSUE TITLE
SDKMAN: JAVA_HOME not set and cannot find javac to deduce location, please set JAVA_HOME

### ISSUE THREAD

**Issue by alu0100207385** [sdkman/sdkman-cli#431]
Hi all,

I'm using sdkman to install groovy. I've followed install instructions from http://sdkman.io/install.html
When I execute "/home/aaron/.sdkman/bin/sdkman-init.sh" I got:

> SDKMAN: JAVA_HOME not set and cannot find javac to deduce location, please set JAVA_HOME

Any ideas? Thanks!

I'm using Ubuntu 14.04LTS (64)

Cheers

**Comment by alu0100207385** [comment]
Ups, I haven't got defined the variable. I have not said anything.

**Comment by marc0der** [comment]
Also remember, do not _execute_ the `sdkman-init.sh` script but _source_ it.

**Comment by YoruNoTori** [comment]
I know  this issue is closed but it would be good to specify some sort of solutions

This could happen to anyone who´s starting to use gradle or any other project, declaring a environment variable is something common to do in console.

Just be sure to install GIT BASH on your machine so you can use UNIX commands on your windows machine

then you just write

`$export JAVA_HOME="/your_java_sdk_directory"`


thats it

**Comment by szcc** [comment]
I add JAVA_HOME to my java.config file under JAVA_CONF="/etc/java/java.conf"
JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.141-1.b16.el7_3.x86_64"
but when I run my grails "run-app" I got error "/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.141-1.b16.el7_3.x86_64" is not a directory.
I try "update-alternatives --display java 
and it showed me the directory /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.141-1.b16.el7_3.x86_64/bin/java

Where is my javac???

**Comment by marc0der** [comment]
@szcc try installing Java with SDKMAN. SDKMAN will dynamically set the `JAVA_HOME` environment variable for you and also allow you to switch between version just like all the other candidates.

@YoruNoTori sorry for not seeing and replying your message. SDKMAN can also be used to install and manage Java, so no need to export `JAVA_HOME`. All it takes is `sdk install java` and the rest is taken care of on your behalf.

**Comment by szcc** [comment]
Thanks for answering my question. I installed java by "sdk install java" yesterday but the JAVA_HOME did not setup by sdkman. when I try to run "grails run-app" I got error java home not found. Then I edited java.config file under JAVA_CONF="/etc/java/java.conf" and JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.141-1.b16.el7_3.x86_64". 
rerun my grails "run-app" I got "not a directory" error.
Now I have to work around it. install sdk 1.8 and set up java home to point it. The idea to use sdkman is manage grails-groovy and java together.....

**Comment by marc0der** [comment]
Please do reach out on our gitter user-issues channel where we can help
you. This is not a bug but a local configuration issue on your machine.

**Comment by szcc** [comment]
thank you. will do

**Comment by vherasme** [comment]
Hi. I am facing this issue when trying to setup a build system with sublime test. I am using **SDKMAN 5.7.3+337**

**Comment by fdchiu** [comment]
I had exactly the same issue. After install JDK 1.7, JAVA_HOME is still pointing to 1.8 which I installed manually through oracle's .dmg file on Mac. sdkman seems not doing what it is claiming to do. It just installed a version of java to: .sdkman/candidates/java directory, but the OS is still pointing to 1.8 which is located at /Library/Java/VirualMachines .

**Comment by minhtran83** [comment]
@fdchiu you can easily set JAVA_HOME like below

`export JAVA_HOME=${SDKMAN_CANDIDATES_DIR}/java/${CURRENT}`

**Comment by marc0der** [comment]
@minhtran83 SDKMAN is already responsible for setting the JAVA_HOME, so you shouldn't be doing it yourself. The problem here is most probably to do with the type of shell being used (login vs interactive), which can lead to inconsistencies in shell state.

**Comment by clehene** [comment]
@marc0der it is?
```
(⎈ |dev-1:default)➜  ~ echo $JAVA_HOME

```

**Comment by marc0der** [comment]
@clehene there really isn't any magic here. You can simply use a text editor to read what `sdkman-init.sh` does when it is sourced.

As stated in the previous comment, most likely you are experiencing this due to a login shell being used, as opposed to an interactive shell.

Also worth noting that most users aren't experiencing what you are, so this is most probably due to an issue in your own environment.

**Comment by joshuapinter** [comment]
> @minhtran83 SDKMAN is already responsible for setting the JAVA_HOME, so you shouldn't be doing it yourself. 

This is bang on. Thanks for clarifying that. I didn't know this and we were unnecessarily setting `JAVA_HOME` on all of our machines. 

SDKMAN FTW!!

**Comment by yangzhongru** [comment]
You just need to set `export JAVA_HOME=/Users/{user}/.sdkman/candidates/java/current`

**Comment by jmagana2000** [comment]
This worked for me. 
`export JAVA_HOME=${SDKMAN_DIR}/candidates/java/current`

**Comment by jaymoid** [comment]
As mentioned above, don't manually set the `JAVA_HOME`, sdk man does this for you when you specify the java version to use: 
`sdk use java java_identifier_goes_here`

for example:
```
$ sdk use java 11.0.11.hs-adpt
Setting java version 11.0.11.hs-adpt as default.

Using java version 11.0.11.hs-adpt in this shell.
$ env | grep JAVA_HOME
JAVA_HOME=/home/james/.sdkman/candidates/java/11.0.11.hs-adpt
```

**Comment by lalik77** [comment]
Hello everybody , can someone help me with this , before installing SDKMAN, i had my local version of JDK . After installing SDKMAN , when i wanted to point sdkman to my local Java  version  , information from this [ <cmd-k>article</cmd-k> ](https://dzone.com/articles/how-to-install-multiple-versions-of-java-on-the-sa)  
====>

`$ ln -s /Library/Java/JavaVirtualMachines/<java-folder>  ~/.sdkman/candidates/java/<java-folder>`

`$ sudo ln -s /Library/Java/JavaVirtualMachines/<java-folder>/Contents/Home/bin  ~/.sdkman/candidates/java/<java-folder>/bin`


then the command on second line give me an error 
->  `ln: /Users/username/.sdkman/candidates/java/jdk1.8.0_271.jdk/bin: No such file or directory ` 

`sdk list java ` give me the next -> at the bottom i have my local version 

<img width="989" alt="Снимок экрана 2021-11-16 в 14 20 08" src="https://user-images.githubusercontent.com/45859153/141976418-b676093b-133f-4605-8a99-b40706a379f4.png">

After that i tell to sdk to use my local version .

<img width="1104" alt="Снимок экрана 2021-11-16 в 14 18 22" src="https://user-images.githubusercontent.com/45859153/141976658-16805df1-639a-4e5d-b05a-f178a4013943.png">

What i am doing wrong  ?  
When a switch back to java installed with sdkman in my case "java versio 16..." , all seems be ok . 
<img width="1131" alt="Снимок экрана 2021-11-16 в 14 30 41" src="https://user-images.githubusercontent.com/45859153/141977796-3d360a3d-bcdb-4482-a5f8-309b117658bd.png">


How finally point SDK MAN to my local version of JAVA ?

**Comment by leandrotapia** [comment]
In my case, after executing `sdk install java <some_version>` it didn't automatically create the `JAVA_HOME` (`env | grep JAVA_HOME` didn't show anything). But executing `sdk use java <some_version>` did the trick. 🙌 
I couldn't figure out what the problem was, so maybe some local configuration made "noise". 🤷‍♂️  

What I had it was java and maven working ok, both were installed with `brew`.
Then I uninstalled java (with `brew`) and (of course) maven couldn't find the `JAVA_HOME`.
Lastly, I installed java with sdkman, and then it happened what I said in the previous paragraph above.

**Comment by mckenzm** [comment]
It's 2023 and this still happens. No JAVA HOME.
Clean install of SDKMAN
Installs Grails OK
sdk install java 8.0.382-zulu  
(no other java installed, so I would not expect update-alternatives issues.  It should set this to the default and do sdk use automatically?

**Comment by mjburghoffer** [comment]
Yeah the issue is you have to do `sdk use java <exact-version-number>`, and it will set `JAVA_HOME` for you.  Sadly, there is no `sdk use java` or `sdk use java current` or even `sdk use java "$(sdk current java)"`.

Here is the line of bash that I found to actually work, since this tool obviously doesn't want to be *too* friendly to use out-of-the-box:

`sdk use java "$(sdk current java | tail -1 | awk -F'. ' ' { print $NF } ')"`

**Comment by NoPhaseNoKill** [comment]
> @clehene there really isn't any magic here. You can simply use a text editor to read what `sdkman-init.sh` does when it is sourced.
> 
> As stated in the previous comment, most likely you are experiencing this due to a login shell being used, as opposed to an interactive shell.
> 
> Also worth noting that most users aren't experiencing what you are, so this is most probably due to an issue in your own environment.

This is contrary to what I'm seeing right now when facing the issue.

Taken from here: https://unix.stackexchange.com/questions/26676/how-to-check-if-a-shell-is-login-interactive-batch (261 upvotes, so the assumption I'm making here is that this how to 'properly' check it).

Running both commands gives me:

$ [[ $- == *i* ]] && echo 'Interactive' || echo 'Not interactive'
Interactive

shopt -q login_shell && echo 'Login shell' || echo 'Not login shell'
Not login shell

My .bashrc file contains the below right at the end of it:

```
#THIS MUST BE AT THE END OF THE FILE FOR SDKMAN TO WORK!!!
export SDKMAN_DIR="$HOME/.sdkman"
[[ -s "$HOME/.sdkman/bin/sdkman-init.sh" ]] && source "$HOME/.sdkman/bin/sdkman-init.sh"
```
In the interest of actually providing some feedback and not just complaining (otherwise we'll never get a resolution to these types of things), I did a little more digging, because I thought it was strange that I was sometimes able to use the command 'java --version' completely fine, and other times it wasn't.

Sometimes I noticed it would come back with: 
```
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.")
```
1. When attempting to run this manually, aka: "$HOME/.sdkman/bin/sdkman-init.sh" <----- bash, I got error: 
```
/$MY_HOME/.sdkman/bin/sdkman-init.sh: Permission denied 
```
Note: '$MY_HOME' is literally a drop in replacement that I've chosen to exclude for this purpose, but is the correct location for what I would expect

2. Checking the file permissions confirmed that this was the expected message given what they were, aka: 
```
-rw-r--r--
```
3. So I then ran: 
```
chmod +x $MY_HOME/.sdkman/bin/sdkman-init.sh
```
4. Checked for my java version inside of the same shell ("java --version"), which once again returned the expected error: 

```
$java --version
Unrecognized option: --version
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit."
```

5. Refreshed my .bashrc file using: "source ~/.bashrc", in case it would in any way shape or form hold any reference to the old permissions

6. Ran "sdk use java 17.0.9-zulu", which returned 

```
Using java version 17.0.9-zulu in this shell.
```

7. Changed it to my default using: "sdk default java 17.0.9-zulu" , which returned 

```setting java 17.0.9-zulu as the default version for all shells.```

8. Ran "java --version" again, to confirm it was set. It was, as it returns: 
```
openjdk 17.0.9 2023-10-17 LTS
OpenJDK Runtime Environment Zulu17.46+19-CA (build 17.0.9+8-LTS)
OpenJDK 64-Bit Server VM Zulu17.46+19-CA (build 17.0.9+8-LTS, mixed mode, sharing)
```

9. Opened a completely new terminal (the previous one was literally the only one open), ran "java --version" and voila - it finally works with:
```
openjdk 17.0.9 2023-10-17 LTS
OpenJDK Runtime Environment Zulu17.46+19-CA (build 17.0.9+8-LTS)
OpenJDK 64-Bit Server VM Zulu17.46+19-CA (build 17.0.9+8-LTS, mixed mode, sharing)
```

10. Closed down all shells, re-opened one, "java --version" again, same thing -> still works. 

11. Opened secondary shell, tried again, still works.

12. In primary shell, used "sdk default java 11.0.21-zulu" -> returned 
```
setting java 11.0.21-zulu as the default version for all shells."
```

13. Went back to secondary shell, did "java --version" again, and voila, it's now working (when previously it was not): 

```
openjdk 11.0.21 2023-10-17 LTS
OpenJDK Runtime Environment Zulu11.68+17-CA (build 11.0.21+9-LTS)
OpenJDK 64-Bit Server VM Zulu11.68+17-CA (build 11.0.21+9-LTS, mixed mode)
```

14. Closed shells again, re-opened, did one more "java --version" -> got expected version 11

```
openjdk 11.0.21 2023-10-17 LTS
OpenJDK Runtime Environment Zulu11.68+17-CA (build 11.0.21+9-LTS)
OpenJDK 64-Bit Server VM Zulu11.68+17-CA (build 11.0.21+9-LTS, mixed mode)
```

My follow up questions to @marc0der:

1. When searching for chmod in the sdkman-init.sh file, it returns no results (aka: "E486: Pattern not found: chmod")
2. Because this init script is running inside of .bashrc, do you think it's reasonable to firstly check if the file has executable permissions, and warn/error if it doesn't, and explain how to do this? It may also need some cache/lock/async if you do choose to do it, because it's probs not worth checking each and every time (particularly when it's inside of .bashrc)
3. The webpage also has no mention of needing to set the file up with executable permissions
4. And neither does the running: "curl -s "https://get.sdkman.io" | bash" from memory (but could be wrong on this point)

tldr; for anyone stumbling across this:

1. ensure you've got relevant executable permissions on the sdkman-init.sh file
2. fix them with something similar to (don't use this blindly, please know what it does first, as you may want to achieve the same thing via other means)
```
 chmod +x $MY_HOME/.sdkman/bin/sdkman-init.sh
```
3. Refresh/restart shells
4. Using the latest version of sdkman, you should be able to then set your default with something like this, and it should now work:

```
sdk default java 17.0.9-zulu
```

**Comment by bouguern** [comment]
> export JAVA_HOME=${SDKMAN_DIR}/candidates/java/current

This is worked for me

**Comment by straiforos** [comment]
I found after installing Java 17 Correto, I could not run mvn. the use command solved this for me. sdk use java 17.0.11-amzn set my JAVA_HOME path. Hope this helps someone else.

**Comment by pavinduLakshan** [comment]
> In my case, after executing `sdk install java <some_version>` it didn't automatically create the `JAVA_HOME` (`env | grep JAVA_HOME` didn't show anything). But executing `sdk use java <some_version>` did the trick. 🙌 I couldn't figure out what the problem was, so maybe some local configuration made "noise". 🤷‍♂️
> 
> What I had it was java and maven working ok, both were installed with `brew`. Then I uninstalled java (with `brew`) and (of course) maven couldn't find the `JAVA_HOME`. Lastly, I installed java with sdkman, and then it happened what I said in the previous paragraph above.

Thanks. saved my day!

**Comment by Jesper-Hustad** [comment]
> In my case, after executing `sdk install java <some_version>` it didn't automatically create the `JAVA_HOME` (`env | grep JAVA_HOME` didn't show anything). But executing `sdk use java <some_version>` did the trick. 🙌 I couldn't figure out what the problem was, so maybe some local configuration made "noise". 🤷‍♂️
> 
> What I had it was java and maven working ok, both were installed with `brew`. Then I uninstalled java (with `brew`) and (of course) maven couldn't find the `JAVA_HOME`. Lastly, I installed java with sdkman, and then it happened what I said in the previous paragraph above.

This worked for me

```
# will output version
sdk current

sdk use [version here]

# test if it worked
echo $JAVA_HOME
```

**Comment by EstopaceMA** [comment]
I've experienced this kind of issue as well, I've installed java before and on my .zshrc there's a JAVA_HOME with a value there, after removing that, SDK man already works fine.

so if you echoed $JAVA_HOME on cli, it will return the directory were the java version is set using sdk man

```
> echo $JAVA_HOME

/Users/mestopace/.sdkman/candidates/java/current
```
